content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+
+---
+
+## 2. Design and Implementation
+
+The implementation consists of several interconnected components that work together to perform effective global value numbering. The overall pipeline is:
+
+```
+Input Program
+    ↓
+SSA Conversion (to_ssa.py with Pruned SSA)
+    ↓
+GVN Pass (gvn.py - dominator tree walk)
+    ↓
+SSA Copy Propagation (in SSA form)
+    ↓
+Dominator-Tree Copy Elimination (in SSA form)
+    ↓
+SSA Lowering (from_ssa.py with fixed copy sequentialization)
+    ↓
+Dead Code Elimination (tdce.py)
+    ↓
+Optimized Program
+```
+
+### 2.1 GVN Algorithm {#21-gvn-algorithm}
+
+#### 2.1.1 Core Data Structures
+
+The GVN implementation (`gvn.py`) uses several key data structures:
+
+**Value Numbers (VN)**: Immutable identifiers assigned to values. Two expressions receive the same VN if and only if they compute the same value.
+
+```python
+@dataclass(frozen=True)
+class VN:
+    n: int
+```
+
+**GVNContext**: Maintains the state during GVN traversal:
+
+```python
+class GVNContext:
+    def __init__(self):
+        self.vn_counter = 0                    # Next VN to allocate
+        self.expr2vn_stack: List[Dict] = []    # Scoped expression → VN mapping
+        self.vn_of: Dict[str, VN] = {}         # SSA name → VN (global)
+```
+
+The expression table uses a **stack of scopes** to handle the dominator tree traversal. When we descend into a dominated block, we push a new scope. Expressions found in dominated blocks are visible to further descendants but not to siblings or ancestors.
+
+#### 2.1.2 Expression Keys
+
+Expressions are canonicalized into hashable keys for the expression table:
+
+```python
+def expr_key(instr: Dict, arg_vns: List[int]) -> Optional[Tuple]:
+    op = instr.get("op")
+    ty = type_key(instr.get("type"))
+
+    if op == "const":
+        return ("const", ty, instr.get("value"))
+
+    # Normalize gt/ge → lt/le with swapped args
+    if op in REL_FLIPS:
+        op = REL_FLIPS[op]
+        arg_vns = list(reversed(arg_vns))
+
+    if op in PURE_BIN or op in PURE_UN:
+        key_args = arg_vns[:]
+        # Canonicalize commutative ops
+        if op in COMMUTATIVE and len(key_args) == 2:
+            if key_args[0] > key_args[1]:
+                key_args = [key_args[1], key_args[0]]
+        return (op, ty, *key_args)
+
+    return None
+```
+
+Key design decisions:
+- **Commutative normalization**: For operations like `add`, `mul`, `and`, we sort arguments by VN so that `a + b` and `b + a` produce the same key.
+- **Relational normalization**: We flip `gt`/`ge` to `lt`/`le` by swapping arguments, reducing the number of distinct expression patterns.
+- **Type inclusion**: Types are included in the key because operations on different types (e.g., `int` vs `float`) are not interchangeable.
+
+#### 2.1.3 Dominator Tree Walk
+
+The core GVN algorithm performs a depth-first traversal of the dominator tree:
```

Comment:
Can you please describe your algorithm/implementation with words? You can link to the code if you want to make it easy to browse the actual Python, but the far more valuable thing to do here would be to just discuss the high-level takeaways you want people to pay attention to.